File: src/main/java/io/openbpm/control/uicomponent/bpmnviewer/BpmnViewer.java
/*
 * Copyright (c) Haulmont 2024. All Rights Reserved.
 * Use is subject to license terms.
 */

package io.openbpm.control.uicomponent.bpmnviewer;


import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.HasElement;
import com.vaadin.flow.function.SerializableConsumer;
import elemental.json.JsonValue;
import io.openbpm.control.uicomponent.bpmnviewer.command.AddMarkerCmd;
import io.openbpm.control.uicomponent.bpmnviewer.command.RemoveMarkerCmd;
import io.openbpm.control.uicomponent.bpmnviewer.command.SetElementColorCmd;
import io.openbpm.control.uicomponent.bpmnviewer.command.SetIncidentCountCmd;
import io.openbpm.control.uicomponent.bpmnviewer.event.SelectionChangedEvent;

/**
 * Interface for a BPMN viewer component that allows visualization and interaction with BPMN diagrams.
 * This interface extends {@link HasElement} to integrate with Vaadin Flow.
 */
public interface BpmnViewer extends HasElement {

    /**
     * The name of the BPMN viewer component.
     */
    String NAME = "bpmnViewer";

    /**
     * Sets the BPMN XML for the viewer component and invokes the callback after the BPMN XML is set.
     *
     * @param bpmnXml  The BPMN XML to be set.
     * @param callback A callback to be invoked after the BPMN XML is set.
     */
    void setBpmnXml(String bpmnXml, SerializableConsumer<JsonValue> callback);

    /**
     * Sets the color of a BPMN element using the specified command.
     * <p>
     * Usage example:
     *
     * <pre>
     * bpmnViewer.setElementColor(new SetElementColorCmd(elementId,
     *    "#000", //stroke color
     *    "#c2d5ed" //fill color
     * ));
     * </pre>
     * @param cmd The command to set the element color.
     */
    void setElementColor(SetElementColorCmd cmd);

    /**
     * Adds a marker to a BPMN element using the specified command.
     * <p>
     * A style for the "highlighted" marker is already defined in the viewer component css file.
     * @param addMarkerCmd The command to add a marker to an element.
     */
    void addMarker(AddMarkerCmd addMarkerCmd);

    /**
     * Removes a marker from a BPMN element using the specified command.
     *
     * @param removeMarkerCmd The command to remove a marker from an element.
     */
    void removeMarker(RemoveMarkerCmd removeMarkerCmd);

    /**
     * Adds a listener to handle BPMN diagram element selection changes.
     *
     * @param listener The listener to handle selection changes.
     */
    void addSelectedElementChangedListener(ComponentEventListener<SelectionChangedEvent> listener);

    /**
     * Resets zoom and centralize a diagram position.
     */
    void resetZoom();

    /**
     * Sets an incident count for provided elements
     * @param cmd The command to set an incident count for elements.
     */
    void setIncidentCount(SetIncidentCountCmd cmd);
}

File: src/main/frontend/src/bpmn-modeler/bpmn-viewer.ts
/*
 * Copyright (c) Haulmont 2024. All Rights Reserved.
 * Use is subject to license terms.
 */

// @ts-ignore
import {css, html, LitElement} from 'lit';
// @ts-ignore
import {customElement} from 'lit/decorators.js';
import BpmViewer from "./bpm/BpmViewer";
import selectionChangedEventHandler from "./component/selectionChangedEventHandler";
import {BpmProcessDefinition} from "./component/types";
import {XmlImportCompleteEvent} from "./component/events";
import {BusinessRuleTaskOverlayClickedEvent} from "./component/events";

// @ts-ignore
@customElement('openbpm-control-bpmn-viewer')
// @ts-ignore
class OpenBpmControlBpmViewer extends LitElement {
    private readonly BPM_VIEWER_HOLDER: string = "bpmnViewerHolder";
    private readonly viewer: BpmViewer;
    private bpmnXml: string | undefined;
    private readonly bpmDrawing: any;
    private readonly canvas: any;
    private shadowRoot: any;
    private zoomScroll: any;
    private processDefinitionsJson: string | undefined;
    private overlays: any;
    private elementRegistry:any;

    static styles = css`
        .highlighted:not(.djs-connection) .djs-visual > :nth-child(1) {
            fill: var(--bpmn-running-activity-color) !important;
        }

        .incident-overlay {
            background-color: var(--bpmn-incident-overlay-bg-color);
            color: var(--bpmn-incident-overlay-text-color);
            border-radius: 50%;
            padding: 0.025em 0.35em;
            line-height: var(--lumo-line-height-xs);
            display: inline-block;
            text-align: center;
            vertical-align: middle;
            font-size: 12px;
            font-weight: bold;
            border: var(--bpmn-incident-overlay-border);
        }

        .decision-instance-link-overlay {
            background-color: var(--bpmn-decision-instance-link-overlay-background);
            cursor: pointer;
            display: flex;
            border-radius: 20%;
            justify-content: center;
            align-items: center;
            width: 1.4em;
            height: 1.4em;
        }
    `;

    constructor() {
        super();
        this.viewer = new BpmViewer();
        this.bpmDrawing = this.viewer.get("bpmDrawing");
        this.canvas = this.viewer.get("canvas");
        this.zoomScroll = this.viewer.get("zoomScroll");
        this.overlays = this.viewer.get("overlays");
        this.elementRegistry = this.viewer.get("elementRegistry");

        let viewer = this.viewer as any;
        const canvas = viewer.get("canvas");
        const element = this;

        viewer.on("selection.changed", (e: any) => selectionChangedEventHandler(element as any, canvas.getRootElement(), e as any));
        viewer.on("import.parse.complete", (e: any) => {
            const rootElements = e.definitions?.rootElements as Array<any> || [];
            const processDefinitions: BpmProcessDefinition[] = [];
            rootElements.forEach((item: any) => {
                if (item.$type === "bpmn:Process") {
                    processDefinitions.push({
                        key: item.id,
                        name: item.name
                    });
                }
            });
            this.processDefinitionsJson = JSON.stringify(processDefinitions);
            this.dispatchEvent(new XmlImportCompleteEvent(this.processDefinitionsJson));
        });
    }

    static get properties() {
        return {
            bpmnXml: {type: String}
        }
    }

    public async reloadSchema(xmlSchema: string) {
        await this.viewer.importXML(xmlSchema)
    }

    // @ts-ignore
    public async getXmlSchema(): Promise<string> {
        const requestXml = await this.viewer.saveXML({
            format: true
        });
        return requestXml.xml!;
    }

    render() {
        return html`
            <div id="${(this.BPM_VIEWER_HOLDER)}" style="width: 100%; height: 100%"/>
        `;
    }

    updated(updatedProps: any) {
        this.awaitRun(() => this.initViewer());
    }

    public setElementColor(cmdJson: any) {
        const cmd = JSON.parse(cmdJson);
        this.awaitRun(() => this.bpmDrawing.setElementColor(cmd));
    }

    public setIncidentCount(cmdJson: any) {
        const cmd = JSON.parse(cmdJson);
        const elements = cmd.elements as Array<{ elementId: string, incidentCount: number, tooltipMessage: string }>;
        this.awaitRun(() => elements.forEach(value => {
            this.overlays.add(value.elementId, {
                html: `<div class="incident-overlay" title="${value.tooltipMessage}">${value.incidentCount}</div>`,
                position: {
                    right: 10,
                    bottom: 15
                }
            })
        }));
    }

    public showBusinessRuleTaskOverlay(cmdJson: any) {
        const cmd = JSON.parse(cmdJson);
        let elements = this.elementRegistry.getAll();
        for (let i = 0; i < elements.length; i++) {
            let value = elements[i];
            if (value.type == "bpmn:BusinessRuleTask" && value.id == cmd.activityId) {
                const htmlDiv = document.createElement('div');
                htmlDiv.innerHTML = `
                <div class="decision-instance-link-overlay" title="${cmd.tooltipMessage}">
                    <svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
                        <path d="M16 7.9L10 3V6C9.5 6 8.9 6 8 6C0 6 0 14 0 14C0 14 1 10 7.8 10C8.9 10 9.6 10 10 10V12.9L16 7.9Z" fill="#ffffff"/>
                    </svg>
                </div>
                `;
                (function(element:OpenBpmControlBpmViewer, decisionInstanceId:string) {
                    htmlDiv.addEventListener('click', (event:MouseEvent) => {
                        element.dispatchEvent(new BusinessRuleTaskOverlayClickedEvent(decisionInstanceId));
                    });
                })(this, cmd.decisionInstanceId);
                this.awaitRun(() =>
                    this.overlays.add(value.id, {
                        html: htmlDiv,
                        position: {
                            left: -10,
                            bottom: 15
                        }
                    }
                ));
                break;
            }
        };
    }

    public addMarker(cmdJson: any) {
        const cmd = JSON.parse(cmdJson);
        this.awaitRun(() => this.canvas.addMarker(cmd.elementId, cmd.marker));
    }

    public removeMarker(cmdJson: any) {
        const cmd = JSON.parse(cmdJson);
        this.awaitRun(() => this.canvas.removeMarker(cmd.elementId, cmd.marker));
    }

    public resetZoom() {
        this.awaitRun(() => this.zoomScroll.reset());
    }

    private initViewer() {
        this.viewer.attachTo(this.shadowRoot.getElementById(this.BPM_VIEWER_HOLDER)!);
        if (this.bpmnXml) {
            this.viewer.importXML(this.bpmnXml).then(r => {
                console.log(r.warnings);
            });
        }
    }

    private awaitRun(callable: () => void) {
        setTimeout(() => {
            try {
                callable();
            } catch (err) {
                console.log('error while running callable', err);
            }
        }, 100);
    }
}
